packages/core-bridge/ts/worker-tuner.ts

```diff
@@ -69,3 +87,102 @@ export interface FixedSizeSlotSupplier {
   // The maximum number of slots that can be issued
   numSlots: number;
 }
+
+/**
+ * The interface can be implemented to provide custom slot supplier behavior.
+ */
+export interface CustomSlotSupplier<SI extends SlotInfo> {
+  type: 'custom';
+
+  /**
+   * This function is called before polling for new tasks. Your implementation should block until a
+   * slot is available then return a permit to use that slot.
+   *
+   * The only acceptable exception to throw is AbortError, any other exceptions thrown will be
+   * logged and ignored.
+   *
+   * @param ctx The context for slot reservation.
+   * @param abortSignal The SDK may decide to abort the reservation request if it's no longer
+   *   needed. Implementations may clean up and then must reject the promise with AbortError.
+   * @returns A permit to use the slot which may be populated with your own data.
+   */
+  reserveSlot(ctx: SlotReserveContext, abortSignal: AbortSignal): Promise<SlotPermit>;
+
+  /**
+   * This function is called when trying to reserve slots for "eager" workflow and activity tasks.
+   * Eager tasks are those which are returned as a result of completing a workflow task, rather than
+   * from polling. Your implementation must not block, and If a slot is available, return a permit
+   * to use that slot.
+   *
+   * @param ctx The context for slot reservation.
+   * @returns Maybe a permit to use the slot which may be populated with your own data.
+   */
+  tryReserveSlot(ctx: SlotReserveContext): SlotPermit | undefined;
```

Comment:
I think in the JS world it's a bit clearer to ask a user to explicitly return null instead of undefined if it's by intention like in this case.

Reply:
I can never keep this straight. It's almost like having two kinds of nothing is a bad idea.